File: mentor-matching-scorer/src/main/scala/com/example/scorer/scoring/Constraints.scala
package com.example.scorer.scoring

import com.example.scorer.model.*

object Constraints:

  def passesAll(mentor: MentorProfile, mentee: MenteeProfile): Boolean =
    hasCapacity(mentor) && notInReportingChain(mentor, mentee)

  def hasCapacity(mentor: MentorProfile): Boolean =
    mentor.currentMenteeCount < mentor.capacity

  def notInReportingChain(
    mentor: MentorProfile,
    mentee: MenteeProfile
  ): Boolean =
    !mentee.reportingChain.contains(mentor.userId)

  def filterMentors(
    mentors: List[MentorProfile],
    mentee: MenteeProfile
  ): List[MentorProfile] =
    mentors.filter(passesAll(_, mentee))

File: mentor-matching-orchestrator/src/orchestrator/constraints.ts
import type { RankedMentor, ScoreMenteeProfile } from "../types/index.js";

export interface EnrichedMentor {
  userId: string;
  timezone: string;
  languages: string[];
  skills: string[];
  yearsExperience: number;
  capacity: number;
  currentMenteeCount: number;
  bioText: string | null;
}

export function filterEligibleMentors(
  mentors: EnrichedMentor[],
  reportingChain: string[]
): EnrichedMentor[] {
  return mentors.filter((m) => {
    if (m.currentMenteeCount >= m.capacity) return false;
    if (reportingChain.includes(m.userId)) return false;
    return true;
  });
}

export function safetyFilter(
  rankings: RankedMentor[],
  mentee: ScoreMenteeProfile,
  mentors: EnrichedMentor[]
): RankedMentor[] {
  const mentorMap = new Map(mentors.map((m) => [m.userId, m]));

  return rankings.filter((r) => {
    const mentor = mentorMap.get(r.mentorUserId);
    if (!mentor) return false;
    if (mentor.currentMenteeCount >= mentor.capacity) return false;
    if (mentee.reportingChain.includes(mentor.userId)) return false;
    return true;
  });
}
